### DISCUSSION TITLE
MDF plotting artefacts/problems

### DISCUSSION THREAD

**Discussion by orlandinebsd** [mtex-toolbox/mtex#1204] [Ask Anything]
Hello, all! 

Thanks to Zach Michael's help in my last post, I am pleased to have an additional problem: MDFs! 

When I use this code:
```
[grains,ebsdS.grainId,ebsdS.mis2mean] = calcGrains(ebsdS,'threshold',[1*degree,10*degree],'unitCell','boundary','tight');
gb_allqz = mergedGrains.boundary('quartz','quartz');
mdf_grains_uncor = calcMDF(mergedGrains('quartz').meanOrientation,mergedGrains('quartz').meanOrientation);
mdf_grains_neighbors = calcMDF(gb_allqz.misorientation,gb_allqz.misorientation,'halfwidth',15*degree);
mdf_sub = calcMDF(gb_qzsub.misorientation,gb_qzsub.misorientation);
```

What I _think_ I am doing is: 
- creating an MDF of all the mean grain orientations in my map, including neighboring grains
- creating an MDF of pixel pair misorientations across high-angle grain boundaries (greater than 10 degrees and that have had their Dauphine twins merged at an earlier step, in this case)
- and then creating an MDF of pixel pair misorientations across low-angle grain boundaries (any boundary between 1 and 10 degrees). 

Assuming that I didn't fall all over myself with my assumptions, we get to the two things that I think MTEX actually might need a bit of code change over. 

**First**, when plotting the misorientation angle distributions, what's up with the weird castle-y, crenelation pattern in the MDFs? Surely that is just an artefact that could be smoothed a little for publication purposes? Is this a result of the hardwired 7.5 degree half width in the calcMDF function? I tried overriding that in the mdf_grains_neighbors MDF calculation above, to no apparent effect.

```
figure
plotAngleDistribution(mdf_grains_neighbors,'DisplayName','Correlated','linewidth',1,'Color','blue')
hold on
plotAngleDistribution(mdf_grains_uncor,'DisplayName','Uncorrelated','linewidth',1,'Color','red')
plotAngleDistribution(odf_grains.CS,odf_grains.CS,'displayName','Random','linewidth',2,'Color','black')
hold off
legend('Location','northwest') % update legend
set(gca, 'FontName', 'Times New Roman')
set(gca, 'FontSize', 16)
```

![BH1E quartz grain boundary misorientations](https://user-images.githubusercontent.com/90639180/133305566-c3ca66f8-efbd-4aa9-8f5f-d3040c14a413.png)

**Second**, when plotting the misorientation axes with plotAxisDistribution, it is apparently not possible to use the 'contourf' function? Would that be an easy fix? As much as I enjoy obsessing over colormaps, the contours are probably 50% of the plot's interpretability. 

```
figure
plotAxisDistribution(mdf_sub,'fundamentalRegion','contourf','minmax','antipodal')
crameri lajolla;
annotate(h.symmetrise,'labeled','backgroundColor','w')
mtexColorbar('FontName','Times') 
```

![BH1E subgrain Qz grain boundary misorientation axes mdf](https://user-images.githubusercontent.com/90639180/133306258-1a17e708-ba9d-4a2f-9184-c0be0acff73b.png)

**Sorta related to that**: are the units still multiples of random distribution in this plot as well when I use gb_qzsub.misorientation, which is what the MDF in the last example was based on? How are these two plots of the same data normalizing things differently? Is the 'random distribution' concept (and I will admit that I have _zero_ personal understanding of how that is being calculated) changing between the two versions? 

`plotAxisDistribution(gb_qzsub.misorientation,'fundamentalRegion','contourf','minmax','antipodal')`

![BH1E subgrain Qz grain boundary misorientation axes mdf](https://user-images.githubusercontent.com/90639180/133307771-b8e2f732-a3e6-4653-8a1f-d1537a8dcd81.png)

**Comment by kilir** [comment]
Hi,
with respect to the wiggle artifacts, this has been discussed here recently. Solution: https://github.com/mtex-toolbox/mtex/discussions/928#discussioncomment-559359

Second, calcMDF computes a density estimation, in the first case `mdf_grains_uncor`  it does what you want, if you supply the misorientations to it, you probably need not  repeat the input argument `gb_allqz.misorientation`.

Third, no problems with contouring here:
```
mtexdata titanium
grains = ebsd.calcGrains
plotAxisDistribution(grains.boundary('t','t').misorientation,'contourf')
```

Fourth: Normalization. It's normalized to a uniform distribution. However, since the axis distribution from a uniform mdf is not uniform (due to symmetry), one could normalize it with respect to the symmetry-induced axis distribution and/or also with respect to the texture-induced axis distribution, mentioned here: https://meetingorganizer.copernicus.org/EGU2019/EGU2019-8207-1.pdf
Another approach is not to normalize the axis distribution but the uncorrelated mdf by the texture induced mdf , see Pospiech, Lücke, Sztwiertnia, 1993, Acta metall. mater. v41.

Hope that helps.
Cheers,
Rüdiger

**Comment by orlandinebsd** [comment]
Hello, Rüdiger! 

Yes, that does help! Thank you. I found 
`  res = get_option(varargin,'resolution',0.5*degree);`
on line 24 of calcAngleDistribution.m in the mtex-5.7.0\ODFAnalaysis\@ODF folder. Changing the value from 0.5 to 0.05 does seem to nearly remove the crenelation, but it also now takes a minute or two instead of a few seconds: 
![BH1E quartz grain boundary misorientations](https://user-images.githubusercontent.com/90639180/133463617-14baee09-dc7c-4e69-af51-09f92051af0b.png)

Using 0.1 instead is quite a bit faster, but it's probably worth it to stick with the smaller value: 
![BH1E quartz grain boundary misorientations](https://user-images.githubusercontent.com/90639180/133464570-b3a61b1d-a5a8-41fc-92bb-c2d4d6e5fc94.png)

2. I am not sure I understand, although I guess it doesn't sound critical.
3. Correct, as in my last example. The problem with 'countourf' is when using 'plotAxisDistribution' on an MDF. 
4. I love the idea of normalizing all of these plots to the version of 'random' which is based on symmetry. What a cool abstract- I wish I had seen the talk/poster! So, you are saying that I am accidentally normalizing my plot to the symmetry-based 'random' in the MDF axis-angle plot, which is why the values are much closer to 1?
